### PR TITLE
[NEW] Support for MacBooks Touch Bar

### LINKED ISSUE #918: Add Support for Mac Laptop Touchbar
I think we should make usage of the touchbar. Something as simply as having the list of servers and displaying them so that people can easily switch to that server. 

https://github.com/electron/electron/blob/master/docs/api/touch-bar.md

### PR DESCRIPTION
**Pull request by abicur**
@RocketChat/electron

Closes #918

There are two panels:
1. Formatting panel. This panel is equivalent of markdown toolbar. Touches a panel button emits click event for corresponding button in markdown toolbar.
2. Select server panel. This panel may be based on `TouchBarSegmentedControl` or `TouchBarScrubber`. There is some limits on available space in Touch Bar. If the for displaying of panel elements requires more space than is possible, the entire panel is not displayed. There are two suitable elements in the API for displaying a list of servers. `TouchBarSegmentedControl` and `TouchBarScrubber`. `TouchBarSegmentedControl` is more convenient and allows to set the index of the selected object, but it suffers from the limitations described above. `TouchBarScrubber` is less convenient to use and does not allow to set the index of the selected item. But `TouchBarScrubber` allows you to display more information. 
I decided to use `TouchBarSegmentedControl` if it is possible or there is an opportunity to shorten the names of servers. Otherwise, a `TouchBarScrubber` is used.

I am not very familiar with ElectronJS, so maybe i did something inefficiently but I am ready to refactor if you point out the shortcomings. 

[There is a demonstration video.](https://www.youtube.com/watch?v=L678MwuekV0)

### CODE REVIEW

File: src/scripts/touchBar.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+					host.label = `${ host.label.slice(0, amountOfCharsPerHost + additionChar) }..`;
+				}
+				return host;
+			});
+		}
+		return arr;
+	}
+}
+
+export class FormattingPanel extends EventEmitter {
```
Comment (tassoevan) [inline src/scripts/touchBar.js]: There is no need to extend `EventEmitter`.

File: src/scripts/touchBar.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+			items: [
+				new TouchBarLabel({ label: i18n.__('Formatting') }),
+				...formatButtons,
+			],
+		});
+		return this._touchBarGroup;
+	}
+}
+
+export class TouchBarBuilder extends EventEmitter {
```
Comment (tassoevan) [inline src/scripts/touchBar.js]: There is no need to extend `EventEmitter`.

File: src/scripts/touchBar.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	_setActiveServer() {
+		if (this._isSegmentedControl()) {
+			this.control.selectedIndex = this._getActiveServerIndex();
+		} else {
+			this._update();
+		}
+	}
+
+	_setHostsArray() {
+		this._hosts = Object.keys(servers.hosts).map((key) => ({ label: key.replace(this._patternRemoveSchema, ''), host: key }));
```
[the remark this answers is not in the document]
Comment (abicur) [reply src/scripts/touchBar.js]: Yes you are right. I just missed these properties.

File: src/scripts/touchBar.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+			}
+		}
+		return items;
+	}
+
+	_buildTouchBarSegmentedControl() {
+		this.control = new TouchBarSegmentedControl({
+			segmentStyle: 'separated',
+			selectedIndex: this._getActiveServerIndex(),
+			segments: this._hosts,
```
[the remark this answers is not in the document]
Comment (abicur) [reply src/scripts/touchBar.js]: As far as I understand, `TouchBarScrubber` and `TouchBarSegmentedControl` allow you to display either an icon or a label. In my experience, the default icon often does not change. I thought it was therefore better to display the label.

If we want to display both Icon and label, then may be we must use a group of buttons (TouchBarButton can display both icon and label) but the buttons are not related to each other. Therefore, we need more logic to color the selected server. I can look into it.

**Pull request merged**